tests/mongo-tools/helpers.go

```diff
@@ -82,3 +90,79 @@ func getDatabaseState(t *testing.T, ctx context.Context, db *mongo.Database) map
 
 	return dbState
 }
+
+// compareFiles takes two files and checks if they have a same content.
+// If they don't, it prints a short diff view.
+func compareFiles(t *testing.T, file1, file2 *os.File) {
+	t.Helper()
+	h := sha256.New()
+
+	_, err := io.Copy(h, file1)
+	require.NoError(t, err)
+
+	hash1 := h.Sum(nil)
+
+	h.Reset()
+
+	_, err = io.Copy(h, file2)
+	require.NoError(t, err)
+
+	hash2 := h.Sum(nil)
+
+	// compare hashes of both files
+	if assert.Equal(t, hash1, hash2, "Checksums of following files are different:", file1.Name(), file2.Name()) {
+		return
+	}
+
+	content1, err := io.ReadAll(file1)
```

Comment:
done

Reply:
First and foremost thanks for a wide explanation :pray:

> It is not clear why we specifically check that file exists, given that os.Open would return the same error.

This change was made only to make the output little bit more readable if file does not exist.

> I'm not sure what we do next. It seems that this code tries to compare files' content – but why, given that we already compared hashes exactly to prevent reading the whole files into memory?

This part of code is called only when checksums are different and it was added only to be more verbose about specific differences in files.

> And then why do we rewind only the first file but not the second?

Nice catch, it doesn't fail because it's not even called, as checksums are the same